core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -769,6 +800,75 @@ private SubPlan optimizePlan(SubPlan plan)
         {
             // Re-optimize plan here based on available runtime statistics.
             // Fragments changed due to re-optimization as well as their downstream stages are expected to be assigned new fragment ids.
+            plan = updateStagesPartitioning(plan);
```

Comment:
Let's change signature of ` private SubPlan optimizePlan(SubPlan plan)` to 
`Optional<SubPlan> optimizePlan()`

We now already exploit the fact that current presorted plan is cached on the `planInTopologicalOrder`. With current code structure it is not obvious that `plan` passed as arguement and plan shared in state match each-other.

Reply:
Hmm I don't quite understand. When will we return `Optional.empty()`?

Reply:
I was thinking returning `Optional.empty()` if the plan should not change. It requires some call-flow refactoring. Can be a followup.